### PR TITLE
Introducing layout update menus (aka dropdowns)

### PR DESCRIPTION
**Pull request by etpinard**
gifrecord_2016-07-22_180239

Native update menus are coming to plotly.js :tada: 

Update menus are designed to _expose_ top-level `Plotly` methods to manage on-click updates. Each buttons is declared with a `method` and an `args` field which are then translated in a `Plotly[method]` call. For example,

```js
layout.updatemenus = [{
  buttons: [{
    method: 'restyle',
    args: [ 'marker.size', 'red' ]
  }]
}];

// is on-click translated to
Plotly.restyle(gd, 'marker.size', 'red')
```

I chose to link them to `layout.updatemenus` instead of e.g. `layout.dropdowns` to allow adding support for other menu _views_ (e.g. static buttons) down the road.

TODOs:
- [x] add tests (once I get some :+1: on the attribute set + functionality)
- [x] agree on default styling cc @delekru
- [ ] decide whether or not and which add more attributes
- [x] decide if we should validate the button `args` for a given button `method` or let plotly error out. **Answer:** let plotly error out.

cc @jackparmer @chriddyp

### CODE REVIEW

**Comment by etpinard** [comment]
To reproduce the graph :arrow_double_up: 

```js
function makeTrace(i) {
    return {
        y: Array.apply(null, Array(10)).map(() => Math.random()),
        line: { 
            shape: 'spline' ,
            color: 'red'
        },
        visible: i === 0,
        name: 'Data set ' + i,

    };
}

Plotly.plot(Tabs.fresh(), [0, 1, 2, 3].map(makeTrace), {
    updatemenus: [{
        buttons: [{
            method: 'restyle',
            args: ['line.color', 'red'],
            label: 'red'
        }, {
            method: 'restyle',
            args: ['line.color', 'blue'],
            label: 'blue'
        }, {
            method: 'restyle',
            args: ['line.color', 'green'],
            label: 'green'
        }] 
    }, {
        y: 0.7,
        buttons: [{
            method: 'restyle',
            args: ['visible', [true, false, false, false]],
            label: 'Data set 0'
        }, {
            method: 'restyle',
            args: ['visible', [false, true, false, false]],
            label: 'Data set 1'
        }, {
            method: 'restyle',
            args: ['visible', [false, false, true, false]],
            label: 'Data set 2'
        }, {
            method: 'restyle',
            args: ['visible', [false, false, false, true]],
            label: 'Data set 3'
        }]
    }]
});
```

**Comment by jackparmer** [comment]
Nice!💃 from me once @delekru is happy with the dropdown colors and styling.

**Comment by delekru** [comment]
@etpinard How about something simple like this?

dropdown

I can change the light-blue hover colour to something a bit more neutral if you'd like.

**Comment by etpinard** [comment]
@delekru looks dope. 

The styling constants are [here](https://github.com/plotly/plotly.js/pull/770/files#diff-dadd5b82ed26b97b7851e99d3e2cddaaR50) and :arrow_down:. We would also need to change the default border width [here](https://github.com/plotly/plotly.js/pull/770/files#diff-b0f64641167e353f2219a0d37bb541aeR140).

The headers and buttons are svg `<rect>`.

**Comment by sudburyrob** [comment]
Are we replicating the same style elements as plotly 2.0?

**Comment by etpinard** [comment]
Screenie of the new @delekru -proof style:

image

File: src/components/updatemenus/attributes.js [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    // add more global settings?
+    //
+    // width
+    // height (instead of being inferred by the label dimensions)
+    // title (above header)
+    // header borderwidth
+    // active bgcolor
+    // hover bgcolor
+    // gap between buttons
+    // gap between header & buttons
```
Comment (bpostlethwaite) [inline src/components/updatemenus/attributes.js]: should we move this comment to an issue or link to an issue?

File: src/components/updatemenus/attributes.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+            'Sets the update menu\'s vertical position anchor',
+            'This anchor binds the `y` position to the *top*, *middle*',
+            'or *bottom* of the range selector.'
+        ].join(' ')
+    },
+
+    font: extendFlat({}, fontAttrs, {
+        description: 'Sets the font of the update menu button text.'
+    }),
+
```
Comment (bpostlethwaite) [inline src/components/updatemenus/attributes.js]: :cow2: some of these attributes have a space between them some do not. Non blocking obviously.

File: src/components/updatemenus/defaults.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    for(var i = 0; i < buttonsIn.length; i++) {
+        buttonIn = buttonsIn[i];
+        buttonOut = {};
+
+        if(!Lib.isPlainObject(buttonIn) || !Array.isArray(buttonIn.args)) {
+            continue;
+        }
+
+        // Should we do some validation for 'args' depending on `method`
+        // or just let Plotly[method] error out?
```
Comment (bpostlethwaite) [inline src/components/updatemenus/defaults.js]: link to an issue?

File: src/components/updatemenus/draw.js [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+        d3.select(this).remove();
+
+        gButton
+            .call(removeAllButtons)
+            .attr(constants.menuIndexAttrName, '-1');
+
+        Plots.autoMargin(gd, constants.autoMarginIdRoot + menuOpts._index);
+    });
+
+    // find dimensions before plotting anything (this mutates menuOpts)
```
Comment (bpostlethwaite) [inline src/components/updatemenus/draw.js]: nice commenting in this file :)

**Comment by bpostlethwaite** [comment]
Looks nice and clean! I didn't get deep into the d3 code as I am not strong in the d3 arts but your commenting made it possible to understand what was going on, nice work!

:dancer:

File: src/components/updatemenus/attributes.js (hunk 1 above)
Comment (etpinard) [inline src/components/updatemenus/attributes.js]: Good point. Thanks!

File: src/components/updatemenus/attributes.js (hunk 2 above)
Comment (etpinard) [inline src/components/updatemenus/attributes.js]: Correct. This is on purpose actually. I like to group e.g. position attributes and style attribute together.

File: src/components/updatemenus/attributes.js (hunk 1 above)
Comment (etpinard) [inline src/components/updatemenus/attributes.js]: moved to https://github.com/plotly/plotly.js/issues/810

File: src/components/updatemenus/defaults.js (hunk 3 above)
Comment (etpinard) [inline src/components/updatemenus/defaults.js]: moved to https://github.com/plotly/plotly.js/issues/810

**Pull request merged**